coderd/promoauth/oauth2.go

```diff
@@ -0,0 +1,163 @@
+				Help:      "The total number of api calls made to external oauth2 providers. 'status_code' will be 0 if the request failed with no response.",
+			}, []string{
+				"name",
+				"source",
+				"status_code",
+			}),
+		},
+	}
+}
+
+func (f *Factory) New(name string, under OAuth2Config) *Config {
+	return &Config{
+		name:       name,
+		underlying: under,
+		metrics:    f.metrics,
+	}
+}
+
+type Config struct {
+	// Name is a human friendly name to identify the oauth2 provider. This should be
+	// deterministic from restart to restart, as it is going to be used as a label in
+	// prometheus metrics.
+	name       string
+	underlying OAuth2Config
+	metrics    *metrics
+}
+
+func (c *Config) Do(ctx context.Context, source string, req *http.Request) (*http.Response, error) {
+	cli := c.oauthHTTPClient(ctx, source)
+	return cli.Do(req)
+}
+
+func (c *Config) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption) string {
+	// No external requests are made when constructing the auth code url.
+	return c.underlying.AuthCodeURL(state, opts...)
+}
+
+func (c *Config) Exchange(ctx context.Context, code string, opts ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
+	return c.underlying.Exchange(c.wrapClient(ctx, "Exchange"), code, opts...)
+}
+
+func (c *Config) TokenSource(ctx context.Context, token *oauth2.Token) oauth2.TokenSource {
+	return c.underlying.TokenSource(c.wrapClient(ctx, "TokenSource"), token)
+}
+
+// wrapClient is the only way we can accurately instrument the oauth2 client.
+// This is because method calls to the 'OAuth2Config' interface are not 1:1 with
+// network requests.
+//
+// For example, the 'TokenSource' method will return a token
+// source that will make a network request when the 'Token' method is called on
+// it if the token is expired.
+func (c *Config) wrapClient(ctx context.Context, source string) context.Context {
+	return context.WithValue(ctx, oauth2.HTTPClient, c.oauthHTTPClient(ctx, source))
+}
+
+// oauthHTTPClient returns an http client that will instrument every request made.
+func (c *Config) oauthHTTPClient(ctx context.Context, source string) *http.Client {
+	cli := &http.Client{}
+
+	// Check if the context has a http client already.
+	if hc, ok := ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		cli = hc
+	}
+
+	// The new tripper will instrument every request made by the oauth2 client.
+	cli.Transport = newInstrumentedTripper(c, source, cli.Transport)
+	return cli
+}
+
+type instrumentedTripper struct {
+	c          *Config
+	source     string
+	underlying http.RoundTripper
+}
+
+// newInstrumentedTripper intercepts a http request, and increments the
+// externalRequestCount metric.
+func newInstrumentedTripper(c *Config, source string, under http.RoundTripper) *instrumentedTripper {
+	if under == nil {
+		under = http.DefaultTransport
+	}
+
+	// If the underlying transport is the default, we need to clone it.
+	// We should also clone it if it supports cloning.
```

Comment:
How about testing for `under.(interface{ Clone() ... })` instead? That would cover the "supports cloning" case.

Reply:
I wish I could but the method signature is:
```golang
func (t *Transport) Clone() *Transport
```

So the interface of `Clone() http.RoundTripper` does not match, and the default transport would not implement it.

Reply:
Ah, of course. That's unfortunate :(.